packages/eslint-plugin-turbo/lib/utils/calculate-inputs.ts

```diff
@@ -0,0 +1,449 @@
+    };
+  };
+};
+
+type ProjectTests = {
+  global: EnvironmentTest;
+  globalTasks: {
+    [script: string]: EnvironmentTest;
+  };
+  workspaceTasks: {
+    [workspace: string]: {
+      [script: string]: EnvironmentTest;
+    };
+  };
+};
+
+// Process inputs for `EnvironmentConfig`s
+
+function processLegacyConfig(legacyConfig: string[] | undefined): string[] {
+  if (!legacyConfig) {
+    return [];
+  }
+
+  const processed = legacyConfig
+    // filter for env vars
+    .filter((dep) => dep.startsWith("$"))
+    // remove leading $
+    .map((variable) => variable.slice(1));
+
+  // After processing length is 0, 1, or more than 1.
+  switch (processed.length) {
+    case 0:
+      return [];
+    case 1:
+      return processed;
+    default:
+      return [...new Set(processed)].sort();
+  }
+}
+
+function processEnv(env: string[] | undefined): string[] {
+  if (!env) {
+    return [];
+  }
+
+  switch (env.length) {
+    case 0:
+      return [];
+    case 1:
+      return [env[0]];
+    default:
+      return [...new Set(env)].sort();
+  }
+}
+
+function processPassThroughEnv(
+  passThroughEnv: string[] | null | undefined
+): string[] | null {
+  if (!passThroughEnv) {
+    return null;
+  }
+
+  switch (passThroughEnv.length) {
+    case 0:
+      return [];
+    case 1:
+      return [passThroughEnv[0]];
+    default:
+      return [...new Set(passThroughEnv)].sort();
+  }
+}
+
+function processDotEnv(
+  workspacePath: string,
+  filePaths: string[] | null | undefined
+): DotEnvConfig | null {
+  if (!filePaths) {
+    return null;
+  }
+
+  const hashEntries: [string, string][] = [];
+  filePaths.reduce((accumulator, filePath) => {
+    const hash = crypto.createHash("sha1");
+    try {
+      const fileContents = fs.readFileSync(path.join(workspacePath, filePath));
+      hash.update(fileContents);
+      accumulator.push([filePath, hash.digest("hex")]);
+    } catch (_) {}
+
+    return accumulator;
+  }, hashEntries);
+
+  return {
+    filePaths,
+    hashes: Object.fromEntries(hashEntries),
+  };
+}
+
+// Generate `EnvironmentConfig`s
+
+function processGlobal(
+  workspacePath: string,
+  rootTurboJson: RootSchema
+): EnvironmentConfig {
+  return {
+    legacyConfig: processLegacyConfig(rootTurboJson.globalDependencies),
+    env: processEnv(rootTurboJson.globalEnv),
+    passThroughEnv: processPassThroughEnv(rootTurboJson.globalPassThroughEnv),
+    dotEnv: processDotEnv(workspacePath, rootTurboJson.globalDotEnv),
+  };
+}
+
+function processTask(workspacePath: string, task: Pipeline): EnvironmentConfig {
+  return {
+    legacyConfig: processLegacyConfig(task.dependsOn),
+    env: processEnv(task.env),
+    passThroughEnv: processPassThroughEnv(task.passThroughEnv),
+    dotEnv: processDotEnv(workspacePath, task.dotEnv),
+  };
+}
+
+const TEST_FALSE = (_: string): boolean => false;
+function generateEnvironmentTest(
+  config: EnvironmentConfig,
+  workspacePath: string | undefined
+): EnvironmentTest {
```

Comment:
Given an `EnvironmentConfig` give me an `EnvironmentTest`.

This turns the configuration into actual functions which return `boolean`s.

Each piece has to be individually tested.